### PR TITLE
Rough Draft: upgrade to ipywidgets 7.x.x (widget protocol version 2.0.0)

### PR DESCRIPTION
**Pull request by rehno-lindeque**
This is a very rough experiment I've been doing related to #870.

Will see, but I somewhat doubt I'll be able to get this draft PR to a merge-able place. I felt that perhaps it'll be some help to someone eventually looking at this however.

### CODE REVIEW

**Comment by vaibhavsagar** [comment]
Thanks for looking into this :smile:.

File: ihaskell-display/ihaskell-widgets/src/IHaskell/Display/Widgets/Button.hs [hunk 1]
```diff
@@ -58,7 +61,17 @@ mkButton = do
...
+      [ DisplayData
+          (MimeCustom "application/vnd.jupyter.widget-view+json")
+          (L.toStrict
+            (encodeToLazyText
+              (object
+                [ "model_id" .= uuid b
+                , "version_major" .= (2 :: Integer)
+                , "version_minor" .= (0 :: Integer)
+                ])))
+      ]
```
Comment (rehno-lindeque) [inline ihaskell-display/ihaskell-widgets/src/IHaskell/Display/Widgets/Button.hs]: I'm battling a little bit with getting the order of messages correct with this approach. It appears as though:

```
b <- mkButton
b
```

works ok.

However, for:

```
mkButton
```

the `display_data` message arrives before `comm_open` does.

Do you possibly have a hint as to why? My instinct was that `flushWidgetMessages` would need to preceed https://github.com/gibiansky/IHaskell/blob/c50e5fc15ff7dcf784fc06dcd3f7318b525f8c98/main/Main.hs#L194, but I'm not having much success with that.

Comment (vaibhavsagar) [reply ihaskell-display/ihaskell-widgets/src/IHaskell/Display/Widgets/Button.hs]: I don't have much insight into how `ihaskell-widgets` works, so I unfortunately can't be of much help here.

Comment (rehno-lindeque) [reply ihaskell-display/ihaskell-widgets/src/IHaskell/Display/Widgets/Button.hs]: No worries, thanks

Comment (rehno-lindeque) [reply ihaskell-display/ihaskell-widgets/src/IHaskell/Display/Widgets/Button.hs]: solved with https://github.com/gibiansky/IHaskell/pull/1091/commits/27981f9b20c1562307b1742cc41b7976d1b7545e

**Comment by jamesdbrock** [comment]
Hi @rehno-lindeque , thanks for doing this work. I want to you to know that I just submitted #1145 , which implements message buffers, which the JupyterLab guys on Gitter told me is necessary for ipywidgets Images.

https://github.com/jupyter/jupyter_client/issues/520

https://github.com/gibiansky/IHaskell/blob/master/ihaskell-display/ihaskell-widgets/src/IHaskell/Display/Widgets/Image.hs

**Comment by jamesdbrock** [comment]
I'm trying to continue this work.

**Comment by jamesdbrock** [comment]
It looks like you got the Button widget working @rehno-lindeque .

**Comment by jamesdbrock** [comment]
I'm seeing this though, even though Button works.

```
[E 04:07:00.497 NotebookApp] Notebook JSON is invalid: {'model_id': 'a0c7db62-f3fa-4246-84ac60e0af5668ed', 'version_major': 2, 'version_minor': 0} is not valid under any of the given schemas
    
    Failed validating 'oneOf' in display_data['properties']['data']['patternProperties']['^(?!application/json$)[a-zA-Z0-9]+/[a-zA-Z0-9\\-\\+\\.]+$']:
    
    On instance['cells'][27]['outputs'][0]['data']['application/vnd.jupyter.widget-view+json']:
    {'model_id': 'a0c7db62-f3fa-4246-84ac-60e0af5668ed',
     'version_major': 2,
     'version_minor': 0}
```

**Comment by jamesdbrock** [comment]
It looks like __jupyter-widgets__ is prepping for version 8. https://github.com/jupyter-widgets/ipywidgets/blob/master/packages/schema/jupyterwidgetmodels.v8.md

**Comment by rehno-lindeque** [comment]
Thank you very much for picking this up @jamesdbrock.

I only needed extremely basic jupyter-widgets support for my purposes, so I haven't looked at it past that - but I'm happy to see this being worked on.

(I used a slightly (ahem) "creative" solution to generate elm widgets inside ihaskell notebooks in order to side-step jupyter widgets entirely)

**Comment by jamesdbrock** [comment]
@rehno-lindeque Well I've been trying to build on your work but so far I've gotten exactly nowhere. I'm pretty sure that for starters we should put the Widget protocol `"version: 2.0.0"` in the header metadata only of the Widget comm messages, not of every message, but when I try to do that then it breaks the basic Button widget that you got working on your branch, and the notebook displays the error message

> A Jupyter widget could not be displayed because the widget state could not be found. This could happen if the kernel storing the widget is no longer available, or if the widget state was not saved in the notebook. You may be able to create the widget by running the appropriate cells.

My failures are on this branch, for anyone who cares:

https://github.com/jamesdbrock/IHaskell/commits/widgets-revive-2

**Comment by rehno-lindeque** [comment]
I'm sorry about that @jamesdbrock... Things seem to be working OK for me for the button at least:

button-widget hs
jupyterlab-logs


Quick question: I am running jupyterlab, not jupyter. How likely do you think it might be that this would make a difference? I seem to remember that jupyterlab was diverging from jupyter when I was doing this.

I could possibly supply a reproducible nix environment via [tweag/jupyterWith](https://github.com/tweag/jupyterWith) that uses my branch if that would help?

**Comment by jamesdbrock** [comment]
This PR has been superceded by #1242

**Comment by vaibhavsagar** [comment]
Thanks for working on this @rehno-lindeque!

**Pull request closed without merging**